pilot/helpers/agents/CodeMonkey.py

```diff
@@ -1,38 +1,353 @@
-            for file_data in changes:
-                self.project.save_file(file_data)
+            if len(blocks) % 2 != 0:
+                print(f"Incorrect number of old→new code blocks ({len(blocks)}) when updating file, asking LLM to retry")
 
+                # If updating is more complicated than just replacing the complete file, don't bother.
+                if len(llm_response) > len(file_content):
+                    replace_complete_file = True
+                    break
+
+                llm_response = convo.send_message('utils/llm_response_error.prompt', {
+                    "error": "Each change should contain exactly one old and exactly one block of code."
+                })
+                exchanged_messages += 2
+                continue
+
+            # Replace old code blocks with new code blocks
+            errors = []
+            for i, (old_code, new_code) in enumerate(zip(blocks[::2], blocks[1::2])):
+                try:
+                    old_code, new_code = self.dedent(old_code, new_code)
+                    content = self.replace(content, old_code, new_code)
+                except ValueError as err:
+                    errors.append((i + 1, str(err)))
+
+            if not errors:
+                break
+
+            print(f"{len(errors)} error(s) while trying to update file, asking LLM to retry")
+
+            if len(llm_response) > len(file_content):
+                # If updating is more complicated than just replacing the complete file, don't bother.
+                replace_complete_file = True
+                break
+
+            # Otherwise, identify the problem block(s) and ask the LLM to retry
+            if content != file_content:
+                error_text = (
+                    "Some changes were applied, but these failed:\n" +
+                    "\n".join(f"Error in change {i}:\n{err}" for i, err in errors) +
+                    "\nPlease fix the errors and try again (only output the blocks that failed to update, not all of them)."
+                )
+            else:
+                error_text = "\n".join(f"Error in change {i}:\n{err}" for i, err in errors)
+
+            llm_response = convo.send_message('utils/llm_response_error.prompt', {
+                "error": error_text,
+            })
+            exchanged_messages += 2
+        else:
+            # We failed after a few retries, so let's just replace the complete file
+            print(f"Unable to modify file, asking LLM to output the complete new file")
+            replace_complete_file = True
+
+        if replace_complete_file:
+            content = self.replace_complete_file(
+                convo,
+                standalone,
+                code_changes_description,
+                file_content,
+                file_name, files
+            )
+            blocks = self.get_code_blocks(llm_response)
```

Comment:
why do we call `self.get_code_blocks(llm_response)` ? I don't see where is this `blocks` used

Reply:
This was accidentially left after I refactored that code to call a function (which handles it already). Removed in 8e301fa1cee720bd789e977e630c8adf55676bb1